core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2016 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "wsched/ChunkTaskList.h"
+
+// System headers
+#include "global/Bug.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Qserv headers
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wsched.ChunkTaskList");
+}
+
+namespace lsst {
+namespace qserv {
+namespace wsched {
+
+
+/// Queue a Task with other tasks on the same chunk.
+void ChunkTaskList::queTask(wbase::Task::Ptr const& task) {
+    std::lock_guard<std::mutex> lg(_mapMx);
+    int chunkId = task->getChunkId();
+    auto iter = _chunkMap.find(chunkId);
+    if (iter == _chunkMap.end()) {
+        iter = _insertChunkTask(chunkId);
+    }
+    ++_taskCount;
+    iter->second->queTask(task);
+}
+
+
+/// Insert a new ChunkTask object into the map.
+/// Precondition: _listMapMx must be locked.
+/// @return an iterator from the map pointing to the new object.
+ChunkTaskList::ChunkMap::iterator ChunkTaskList::_insertChunkTask(int chunkId) {
+    std::pair<int, ChunkTasks::Ptr> ele(chunkId, std::make_shared<ChunkTasks>(chunkId, _memMan));
+    auto res = _chunkMap.insert(ele); // insert should fail if the key already exists.
+    if (!res.second) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTaskList::_insertChunkTask entry already existed " << chunkId);
+    }
+    return res.first;
+}
+
+/// @return true if this object is ready to provide a Task from its queue.
+bool ChunkTaskList::ready(bool useFlexibleLock) {
+    std::lock_guard<std::mutex> lock(_mapMx);
+    return _ready(useFlexibleLock);
+}
+
+/// Precondition: _queueMutex must be locked
+/// If true is returned, _readyChunk will point to a chunk with a Task that is ready to run.
+/// This function starts checking at the _activeChunk and only looks to the next chunk
+/// if there are no tasks to run on the current chunk.
+/// The _activeChunk advances when all of its Tasks have completed.
+/// @return true if this object is ready to provide a Task from its queue. _readyChunk != nullptr.
```

Comment:
I was trying to say that if this function returns true, _readyChunk will point to the chunk with a Task to run. If it returns false, _readyChunk will be nullptr. I'll fix the comment.